[quant][graphmode][fx] Add reference option support for linear_static_fp16

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52865 [quant][graphmode][fx][fp16] Add fp16 support for silu
* #52864 [quant][graphmode][fx][fp16] Add fp16 support for tanh
* #52863 [quant][graphmode][fx][fp16] Add fp16 support for sigmoid
* #52862 [quant][graphmode][fx][fp16] Add fp16 support for layer_norm
* #52811 [quant][graphmode][fx][fp16] Add fp16 support for torch.sum
* #52810 [quant][graphmode][fx][fp16] Add fp16 support for div
* #52809 [quant][graphmode][fx][fp16] Add fp16 support for sub
* #52808 [quant][graphmode][fx] Add support for fp16 bmm pattern
* #52807 [quant][graphmode][fx][test][refactor] Refactoring binary op tests to split int8 and float16 tests
* #52714 [quant][graphmode][fx][fp16] Add fp16 support for {add|mul}{_relu}
* #52651 [quant][refactor] Merge add and mul handler
* **#52650 [quant][graphmode][fx] Add reference option support for linear_static_fp16**

Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26599803](https://our.internmc.facebook.com/intern/diff/D26599803)

jerryzh168 committed:
[quant][graphmode][fx] Add reference option support for linear_static_fp16

Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for linear_static_fp16"


Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26599803](https://our.internmc.facebook.com/intern/diff/D26599803)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for linear_static_fp16"


Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26599803](https://our.internmc.facebook.com/intern/diff/D26599803)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -449,7 +451,15 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
                 "supported by Linear "
                 "supported dtype combinations are: {}".format(dtypes, supported_dtypes))
-            return quantizer.quantized_graph.node_copy(node, load_arg(quantized=None))
+            if self.relu_node:
+                op_out = quantizer.quantized_graph.node_copy(self.linear_node, load_arg(quantized=False))
+                relu_args = [op_out]
+                relu_args.extend(load_arg(quantized=False)(self.relu_node.args[1:]))
+                relu_kwargs = load_arg(quantized=False)(self.relu_node.kwargs)
+                return quantizer.quantized_graph.create_node(
+                    "call_function", torch.nn.functional.relu, tuple(relu_args), relu_kwargs)
+            else:
+                return quantizer.quantized_graph.node_copy(node, load_arg(quantized=None))
```
just for my own understanding, this is an improvement which is not related to the title of this PR?  Sounds good, just making sure i'm understanding things correctly

jerryzh168 replied:
yeah, this is unrelated to the PR, more of supporting the case when an unsupported qconfig is configured for linear and make sure we still produce correct model and not error out.

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -520,29 +530,31 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
-                prepack_op = get_linear_prepack_op_for_dtype(weight_dtype)
-                packed_weight = quantizer.quantized_graph.create_node(
-                    'call_function', prepack_op, prepack_args, {})
+                # prepacking weights for static int8 quant and dynamic quant
```
for my own understanding, wondering why the code change is for "non-reference option" (line above), but the PR is adding support for reference option?

jerryzh168 replied:
Oh, reference option is supported through modifying `quantize_node` (see previous PR), actually for fp16 static quant both reference and non-reference option would produce the same result

jerryzh168 replied:
here I just added a few dtype checks to make sure the branch conditions are more explicit.

vkuzo commented on torch/quantization/qconfig.py:
```diff
@@ -66,6 +66,8 @@ def __new__(cls, activation=torch.nn.Identity, weight=torch.nn.Identity):
                                          weight=default_weight_observer)
 float16_dynamic_qconfig = QConfigDynamic(activation=PlaceholderObserver.with_args(dtype=torch.float32),
                                          weight=PlaceholderObserver.with_args(dtype=torch.float16))
+float16_static_qconfig = QConfigDynamic(activation=PlaceholderObserver.with_args(dtype=torch.float16),
```
are we treating fp16 emulation as static quantization?  Thoughts about having it be a new quantization type instead?  cc @raghuramank100

jerryzh168 replied:
yes, I think this falls into the category of static quant, (static, dynamic, weight_only are referring to how we quantize activations)

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for linear_static_fp16"


Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26599803](https://our.internmc.facebook.com/intern/diff/D26599803)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for linear_static_fp16"


Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26599803](https://our.internmc.facebook.com/intern/diff/D26599803)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for linear_static_fp16"


Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26599803](https://our.internmc.facebook.com/intern/diff/D26599803)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for linear_static_fp16"


Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26599803](https://our.internmc.facebook.com/intern/diff/D26599803)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for linear_static_fp16"


Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26599803](https://our.internmc.facebook.com/intern/diff/D26599803)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for linear_static_fp16"


Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26599803](https://our.internmc.facebook.com/intern/diff/D26599803)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for linear_static_fp16"


Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26599803](https://our.internmc.facebook.com/intern/diff/D26599803)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for linear_static_fp16"


Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26599803](https://our.internmc.facebook.com/intern/diff/D26599803)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for linear_static_fp16"


Summary:
linear_dynamic_fp16 has following dtypes for activation, weight, bias, output:
(fp32, fp16, fp32, fp32)

linear_static_fp16 has following dtypes:
(fp16, fp16, fp16, fp16)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26599803](https://our.internmc.facebook.com/intern/diff/D26599803)

[ghstack-poisoned]
